### PR TITLE
[RW-3999][RISK=LOW] Avoid leaving DB entity references in scope in undefined state

### PR DESCRIPTION
**Pull request by jaycarlton**
The CrudRepository save() method has the undesirable property that it alters the state of its input. Consequently, in order to be safe, the value passed in must be either
- be an unnamed temporary, as in the output of a method call: `fooDao.save(makeFooToSave(...));`
- be immediately overwritten: `dbFoo = fooDao.save(dbFoo);`
- removed from scope, either by having the save call used in a return expression or passed into another method: `return fooDao.save(dbFoo);`

In a few cases (for example void-valued functions), I chose to leave a "useless" assignment as the last statement of the function, just to make sure maintainers wouldn't be tempted to reuse the now dead value. So you might see `void doStuff(DbFoo dbFoo) { dbFoo = fooDao.save(dbFoo); }. IntelliJ will laugh at you for this, but it's legal.

There are a lot more files to check, and I don't know of a good automatic way of doing it, so I just kinda fiddled with this while watching tv.

### CODE REVIEW

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java [hunk 1]
```diff
@@ -176,21 +176,21 @@ public void setUp() throws Exception {
...
     cdrVersion.setBigqueryProject(testWorkbenchConfig.bigquery.projectId);
-    cdrVersionDao.save(cdrVersion);
+    cdrVersion = cdrVersionDao.save(cdrVersion);
 
     workspace = new DbWorkspace();
     workspace.setCdrVersion(cdrVersion);
     workspace.setWorkspaceNamespace(NAMESPACE);
     workspace.setFirecloudName(NAME);
-    workspaceDao.save(workspace);
+    workspace = workspaceDao.save(workspace);
```
Comment (freemabd) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java]: nit: can we rename the variable to dbWorkspace?

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java [hunk 2]
```diff
@@ -176,21 +176,21 @@ public void setUp() throws Exception {
...
     stubMockFirecloudGetWorkspace();
     stubMockFirecloudGetWorkspaceAcl();
 
     Gson gson = new Gson();
     cohort = new DbCohort();
     cohort.setWorkspaceId(workspace.getWorkspaceId());
     cohort.setCriteria(gson.toJson(SearchRequests.males()));
-    cohortDao.save(cohort);
+    cohort = cohortDao.save(cohort);
```
Comment (freemabd) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java]: nit: can we rename the variable to dbCohort?

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java [hunk 3]
```diff
@@ -199,23 +199,19 @@ public void setUp() throws Exception {
             .creationTime(new Timestamp(new Date().getTime()))
             .lastModifiedTime(new Timestamp(new Date().getTime()))
             .cohortId(cohort.getCohortId());
-    cohortReviewDao.save(review);
+    review = cohortReviewDao.save(review);
```
Comment (freemabd) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java]: nit: can we rename the variable to dbReview?

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java [hunk 4]
```diff
@@ -355,7 +351,7 @@ public void createCohortReview() throws Exception {
             + "\"ICD10CM\",\"group\":true,\"attributes\":[],\"ancestorData\":false,\"standard\":false,\"conceptId\":1,\"value\":\"C34\"}],"
             + "\"modifiers\":[]}],\"temporal\":false}],\"excludes\":[]}";
     cohortWithoutReview.setCriteria(criteria);
-    cohortDao.save(cohortWithoutReview);
+    cohortWithoutReview = cohortDao.save(cohortWithoutReview);
```
Comment (freemabd) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java]: nit: can we rename the variable to dbCohortWithoutReview?

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java [hunk 5]
```diff
@@ -355,7 +351,7 @@ public void createCohortReview() throws Exception {
...
 
     org.pmiops.workbench.model.CohortReview cohortReview =
```
Comment (freemabd) [inline api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java]: Can we remove this fully qualified name now that we have DbCohortReview?

File: api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceBQTest.java [hunk 6]
```diff
@@ -123,22 +123,22 @@ public void setUp() {
...
-    cdrVersionDao.save(cdrVersion);
+    cdrVersion = cdrVersionDao.save(cdrVersion);
     CdrVersionContext.setCdrVersionNoCheckAuthDomain(cdrVersion);
 
     DbWorkspace workspace = new DbWorkspace();
     workspace.setCdrVersion(cdrVersion);
     workspace.setName("name");
     workspace.setDataAccessLevelEnum(DataAccessLevel.PROTECTED);
-    workspaceDao.save(workspace);
+    workspace = workspaceDao.save(workspace);
```
Comment (freemabd) [inline api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceBQTest.java]: nit: can we rename the variable to dbWorkspace?

File: api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceBQTest.java [hunk 7]
```diff
@@ -123,22 +123,22 @@ public void setUp() {
...
 
     DbCohort cohort = new DbCohort();
     cohort.setWorkspaceId(workspace.getWorkspaceId());
     cohort.setName("males");
     cohort.setType("AOU");
     Gson gson = new Gson();
     cohort.setCriteria(gson.toJson(SearchRequests.males()));
-    cohortDao.save(cohort);
+    cohort = cohortDao.save(cohort);
```
Comment (freemabd) [inline api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceBQTest.java]: nit: can we rename the variable to dbCohort?

File: api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceBQTest.java [hunk 8]
```diff
@@ -153,7 +153,7 @@ public void setUp() {
     cohortReview.setMatchedParticipantCount(3);
     cohortReview.setReviewedCount(2);
     cohortReview.setReviewSize(3);
-    cohortReviewDao.save(cohortReview);
+    cohortReview = cohortReviewDao.save(cohortReview);
```
Comment (freemabd) [inline api/src/bigquerytest/java/org/pmiops/workbench/cohorts/CohortMaterializationServiceBQTest.java]: nit: can we rename the variable to dbCohortReview?

File: api/src/main/java/org/pmiops/workbench/cohorts/CohortCloningService.java [hunk 9]
```diff
@@ -26,16 +26,15 @@
 
   @Transactional
   public DbCohort cloneCohortAndReviews(DbCohort fromCohort, DbWorkspace targetWorkspace) {
-    final DbCohort duplicatedCohort =
-        cohortFactory.duplicateCohort(
-            fromCohort.getName(), targetWorkspace.getCreator(), targetWorkspace, fromCohort);
-    final DbCohort toCohort = cohortDao.save(duplicatedCohort);
+    final DbCohort toCohort =
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cohorts/CohortCloningService.java]: nit: can we rename the variable to toDbCohort?

File: api/src/main/java/org/pmiops/workbench/cohorts/CohortCloningService.java [hunk 10]
```diff
@@ -26,16 +26,15 @@
...
+        cohortDao.save(
+            cohortFactory.duplicateCohort(
+                fromCohort.getName(), targetWorkspace.getCreator(), targetWorkspace, fromCohort));
     copyCohortAnnotations(fromCohort, toCohort);
 
     for (DbCohortReview fromReview : fromCohort.getCohortReviews()) {
-      final DbCohortReview duplicatedReview =
-          cohortFactory.duplicateCohortReview(fromReview, toCohort);
-      final DbCohortReview toReview = cohortReviewDao.save(duplicatedReview);
+      final DbCohortReview toReview =
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/cohorts/CohortCloningService.java]: nit: can we rename the variable to toDbCohortReview?

File: api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java [hunk 11]
```diff
@@ -78,7 +78,13 @@ public void updateCohortEntry(long workspaceId, long userId, long cohortId) {
   public void updateConceptSetEntry(long workspaceId, long userId, long conceptSetId) {
     Timestamp now = new Timestamp(clock.instant().toEpochMilli());
 
-    DbConceptSet conceptSet = conceptSetDao.findOne(conceptSetId);
+    final DbConceptSet conceptSet = conceptSetDao.findOne(conceptSetId);
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java]: nit: can we rename the variable to dbConceptSet?

File: api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java [hunk 12]
```diff
@@ -78,7 +78,13 @@ public void updateCohortEntry(long workspaceId, long userId, long cohortId) {
...
+    userRecentResourceDao.save(makeUserRecentResource(workspaceId, userId, now, conceptSet));
+  }
+
+  @NotNull
+  private DbUserRecentResource makeUserRecentResource(
+      long workspaceId, long userId, Timestamp now, DbConceptSet conceptSet) {
     DbUserRecentResource resource =
```
Comment (freemabd) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserRecentResourceServiceImpl.java]: nit: can we rename the variable to dbUserRecentResource?

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java (hunk 1 above)
Comment (jaycarlton) [reply api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java]: Yes, I had that in this PR but it drowned out the save-related changes (some of which are subtle).

File: api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java (hunk 5 above)
Comment (jaycarlton) [reply api/src/bigquerytest/java/org/pmiops/workbench/api/CohortReviewControllerBQTest.java]: 👍

**Pull request merged**